### ISSUE TITLE
lint: variables with "__" are triggering non-snake-case

### ISSUE THREAD

**Issue by LoveIsGrief** [rust-lang/rust#85778]
https://github.com/rust-lang/rust/blob/8d1e3d3b74fa80ce14af6ec143f061897684582b/compiler/rustc_lint/src/nonstandard_style.rs#L267

In this logic, two adjacent underscores trigger the `non-snake-case` warning. Currently, I use them to discern the types of my variables. CamelCase classes are turned into snake-case e.g `camel_case` and combined with the name `stack_sidebar__step1` or `step1__stack_sidebar`.

This is useful as I'm currently writing UI code with gtk-rs and gtk has multiple button types (push button, radio button, spin button,...). Having everything be 

`button_step1` and `button_spin_step1` would add confusion. `button_spin_step1` could  either be a `spin button` for step1 or a button for `spin_step1`.

**Comment by Enselic** [comment]
I looked into fixing this, but it turns our there are explicit regression tests for making sure double underscore is not allowed:

https://github.com/rust-lang/rust/blob/cf2dff2b1e3fa55fa5415d524200070d0d7aacfe/tests/ui/lint/lint-non-snake-case-functions.rs#L28

And I think that makes sense for several reasons:
* `__` is non-standard Rust style (anecdotally: I have never seen it before, and I've seen a fair amount of Rust code)
* Arguably in most cases where `__` is encountered, it is a typo on part of the user, worth linting for.

For the cases where use of double underscore is desired, I think it is reasonable to require opting out of this lint.

Triage: Closing as works as designed/won't fix.
